### PR TITLE
WIP: use weaklist rather than weakset 

### PR DESCRIPTION
**Pull request by CJ-Wright**
so that we maintain execution order. Otherwise python sets may change which branch gets executed first.

### CODE REVIEW

**Comment by mrocklin** [comment]
There is a cost to depending on a third party package for this.  There are probably also benefits.  I think that the decision on whether or not to include this is based on those costs and benefits.  I don't currently have a good handle on either.

**Comment by CJ-Wright** [comment]
We could write our own, I don't mind either way. I certainly don't want to end up in a `leftpad` situation, but then we have to maintain it. The lib this PR adds is [less than 100 lines](https://github.com/apieum/weakreflist/blob/master/weakreflist/weakreflist.py).

Is the move from set back to list ok?

**Comment by mrocklin** [comment]
I remain concerned that weaklist isn't standard.  I would expect there to be complications in the future.  To justify this I would prefer to see a variety of use cases that required deterministic ordering.  I don't personally see this, but I appreciate that you have a different set of experiences and may see things that I don't here.  I think that this decision needs more discussion before we consider merging anything like this.

**Comment by CJ-Wright** [comment]
Here is an example
```python
In [2]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'red')
(3, 'blue')

In [3]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'red')
(3, 'blue')

In [4]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'red')
(3, 'blue')

In [5]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'blue')
(3, 'green')

In [6]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'red')
(3, 'blue')
```
It seems that we get different results despite running the same code.
In [5] is doing what I'd expect if `s.pluck(1)` was being run before `s.pluck(0)`.

**Comment by CJ-Wright** [comment]
Note that with this PR I get this:
```python
In [3]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'blue')
(3, 'green')

In [4]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'blue')
(3, 'green')

In [5]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'blue')
(3, 'green')

In [6]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'blue')
(3, 'green')

In [7]: s = Stream()
   ...: b = s.pluck(1)
   ...: a = s.pluck(0)
   ...: c = a.zip_latest(b).sink(print)
   ...: z = [(1, 'red'), (2, 'blue'), (3, 'green')]
   ...: for zz in z:
   ...:     s.emit(zz)
   ...:     
(1, 'red')
(2, 'blue')
(3, 'green')
```
The execution is the same every time and in the same order as the stream nodes were created.

**Comment by CJ-Wright** [comment]
For a more concrete example consider an array that gets processed and then normalized by another number by a zip_latest (or combine_latest) and a map. If the execution order is uncertain then the array and the normalization may not be in sync causing the wrong results. We found a similar example with filenames, where the data wouldn't always be matched to the filename that was created, due to the execution order being off.

**Comment by jrmlhermitte** [comment]
can you please summarize with a for loop?
```python
from streamz import Stream

for i in range(5):
    s = Stream()
    b = s.pluck(1)
    a = s.pluck(0)
    c = a.zip_latest(b).sink(print)
    z = [(1, 'red'), (2, 'blue'), (3, 'green')]
    for zz in z:
        s.emit(zz)
```
```
(1, 'red')
(2, 'blue')
(3, 'green')
(1, 'red')
(2, 'blue')
(3, 'green')
(1, 'red')
(2, 'blue')
(3, 'green')
(1, 'red')
(2, 'red')
(3, 'blue')
(1, 'red')
(2, 'blue')
(3, 'green')
```
I agree, this is an issue.

**Comment by CJ-Wright** [comment]
With combine latest:
```python
In [10]: for i in range(5):
    ...:     s = Stream()
    ...:     b = s.pluck(1)
    ...:     a = s.pluck(0)
    ...:     c = a.combine_latest(b, emit_on=a).sink(print)
    ...:     z = [(1, 'red'), (2, 'blue'), (3, 'green')]
    ...:     for zz in z:
    ...:         s.emit(zz)
    ...:         
(2, 'red')
(3, 'blue')
(2, 'red')
(3, 'blue')
(1, 'red')
(2, 'blue')
(3, 'green')
(2, 'red')
(3, 'blue')
(1, 'red')
(2, 'blue')
(3, 'green')
```

**Comment by jrmlhermitte** [comment]
It fixes `combine_latest`:
```python
from streamz import Stream
import time

print("Starting")
for i in range(5):
    print(f"\nTry {i}")
    s = Stream()
    b = s.pluck(1)
    a = s.pluck(0)
    c = a.combine_latest(b, emit_on=a).sink(print)
    z = [(1, 'red'), (2, 'blue'), (3, 'green')]
    for zz in z:
        s.emit(zz)
        time.sleep(.1)
```
before:
```Starting

Try 0
(2, 'red')
(3, 'blue')

Try 1
(2, 'red')
(3, 'blue')

Try 2
(1, 'red')
(2, 'blue')
(3, 'green')

Try 3
(1, 'red')
(2, 'blue')
(3, 'green')

Try 4
(1, 'red')
(2, 'blue')
(3, 'green')

```

After:
```python
Starting

Try 0
(1, 'red')
(2, 'blue')
(3, 'green')

Try 1
(1, 'red')
(2, 'blue')
(3, 'green')

Try 2
(1, 'red')
(2, 'blue')
(3, 'green')

Try 3
(1, 'red')
(2, 'blue')
(3, 'green')

Try 4
(1, 'red')
(2, 'blue')
(3, 'green')
```

**Comment by jrmlhermitte** [comment]
current code we copied is LGPL:
" Everyone is permitted to copy and distribute verbatim copies of this license document, but changing it is not allowed."
I believe that's compatible with BSD?

File: streamz/weakreflist.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# -*- coding: utf8 -*-
+# This is a copy from weakreflist
+# https://stackoverflow.com/questions/677978/weakref-list-in-python
```
Comment (jrmlhermitte) [inline streamz/weakreflist.py]: https://pypi.python.org/pypi/weakreflist

**Comment by mrocklin** [comment]
Two comments:

1.  weakreflist is LGPL.  We can not include it without changing our license as well.
2.  Does this affect performance?

**Comment by mrocklin** [comment]
Ah, I see that @jrmlhermitte beat me to the LGPL punch.

**Comment by mrocklin** [comment]
I am not a lawyer, but I've tended to avoid touching (or even looking at) LGPL code.  You should probably seek guidance from someone with more legal experience.  I suspect that importing weakreflist might be possible (though I'm still somewhat wary about depending on this (but perhaps that is irrational)).

**Comment by CJ-Wright** [comment]
I'm not certain how to deal with the LGPL issues, we could use [WeakOrderedSet](https://stackoverflow.com/a/7829569/5100330)
Do we have current performance metrics?

**Comment by CJ-Wright** [comment]
Actually we *need* to use WeakOrderedSet since lists can contain multiple references to the same thing.

**Comment by jrmlhermitte** [comment]
Fortunately, it seems that all stackoverflow posts are under a very loose license:
https://stackoverflow.com/help/licensing

I understand that post (and links within) meaning we can copy and paste, modify and redistribute at will so long as we keep the stack overflow link. Is that acceptable?

**Comment by CJ-Wright** [comment]
This also supports putting the execution order on the graph, if you use the source node.

**Comment by mrocklin** [comment]
The OrderedWeakSet implemention looks fine to me.  Does it have any performance impact?

No comments on the graph visualization stuff other than that the extent to which this code is untested remains a slight concern.  My confidence that it will be maintainable in the future is not very high.

**Comment by CJ-Wright** [comment]
Speed test:
I ran
```python
from streamz import Stream
from profilehooks import profile

@profile
def test():
    for i in range(100):
        s = Stream()
        b = s.pluck(1)
        a = s.pluck(0)
        c = a.combine_latest(b, emit_on=a)
        z = [(1, 'red'), (2, 'blue'), (3, 'green')] * 100
        for zz in z:
            s.emit(zz)

test()
```
For master I got:
```
*** PROFILER RESULTS ***
test (/home/christopher/.PyCharm2017.2/config/scratches/scratch_128.py:4)
function called 1 times

         3483687 function calls (3303201 primitive calls) in 2.103 seconds

   Ordered by: cumulative time, internal time, call count
   List reduced from 91 to 40 due to restriction <40>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.054    0.054    2.103    2.103 scratch_128.py:4(test)
59952/30000    0.110    0.000    2.014    0.000 core.py:194(emit)
119952/30000    0.258    0.000    1.321    0.000 core.py:183(_emit)
    60000    0.054    0.000    1.080    0.000 core.py:944(update)
89904/59952    0.079    0.000    0.975    0.000 functools.py:744(wrapper)
89904/59952    0.067    0.000    0.851    0.000 gen.py:1259(convert_yielded)
    59952    0.065    0.000    0.797    0.000 gen.py:672(multi)
    60000    0.067    0.000    0.632    0.000 core.py:814(update)
    59952    0.229    0.000    0.605    0.000 gen.py:793(multi_future)
   240039    0.217    0.000    0.470    0.000 _weakrefset.py:58(__iter__)
   120007    0.092    0.000    0.134    0.000 _weakrefset.py:26(__exit__)
    29952    0.016    0.000    0.133    0.000 concurrent.py:259(add_done_callback)
    59952    0.060    0.000    0.127    0.000 gen.py:659(_contains_yieldpoint)
    29952    0.056    0.000    0.117    0.000 gen.py:822(callback)
    89904    0.060    0.000    0.100    0.000 functools.py:706(dispatch)
   120007    0.061    0.000    0.076    0.000 _weakrefset.py:20(__enter__)
    59952    0.027    0.000    0.064    0.000 {built-in method builtins.all}
420078/419816    0.061    0.000    0.063    0.000 {built-in method builtins.isinstance}
    59956    0.028    0.000    0.053    0.000 {built-in method builtins.any}
    59952    0.028    0.000    0.051    0.000 concurrent.py:272(set_result)
   120007    0.043    0.000    0.043    0.000 _weakrefset.py:16(__init__)
    59952    0.042    0.000    0.042    0.000 concurrent.py:168(__init__)
    89904    0.040    0.000    0.040    0.000 weakref.py:393(__getitem__)
    89904    0.023    0.000    0.037    0.000 gen.py:815(<genexpr>)
   120007    0.031    0.000    0.031    0.000 _weakrefset.py:52(_commit_removals)
    59904    0.020    0.000    0.029    0.000 concurrent.py:355(is_future)
    59952    0.020    0.000    0.028    0.000 concurrent.py:335(__del__)
   119952    0.026    0.000    0.026    0.000 core.py:192(<listcomp>)
    89904    0.021    0.000    0.025    0.000 gen.py:668(<genexpr>)
    29952    0.016    0.000    0.025    0.000 concurrent.py:225(result)
    59952    0.024    0.000    0.024    0.000 concurrent.py:321(_set_done)
   150414    0.018    0.000    0.018    0.000 {method 'add' of 'set' objects}
    60000    0.016    0.000    0.016    0.000 {method 'index' of 'tuple' objects}
   150159    0.014    0.000    0.014    0.000 {method 'remove' of 'set' objects}
    89958    0.012    0.000    0.012    0.000 {method 'append' of 'list' objects}
    60007    0.011    0.000    0.011    0.000 {built-in method builtins.getattr}
    60000    0.009    0.000    0.009    0.000 {method 'extend' of 'list' objects}
    29952    0.009    0.000    0.009    0.000 concurrent.py:219(_clear_tb_log)
    59952    0.008    0.000    0.008    0.000 {built-in method sys.is_finalizing}
      300    0.000    0.000    0.007    0.000 core.py:142(wrapped)
```
With this branch I got:
```
*** PROFILER RESULTS ***
test (/home/christopher/.PyCharm2017.2/config/scratches/scratch_128.py:4)
function called 1 times

         3727756 function calls (3547097 primitive calls) in 2.178 seconds

   Ordered by: cumulative time, internal time, call count
   List reduced from 98 to 40 due to restriction <40>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.054    0.054    2.178    2.178 scratch_128.py:4(test)
60000/30000    0.110    0.000    2.087    0.000 core.py:194(emit)
120000/30000    0.260    0.000    1.409    0.000 core.py:183(_emit)
    60000    0.056    0.000    1.139    0.000 core.py:945(update)
90000/60000    0.080    0.000    0.953    0.000 functools.py:744(wrapper)
90000/60000    0.068    0.000    0.832    0.000 gen.py:1259(convert_yielded)
    60000    0.063    0.000    0.778    0.000 gen.py:672(multi)
    60000    0.064    0.000    0.643    0.000 core.py:815(update)
    60000    0.222    0.000    0.587    0.000 gen.py:793(multi_future)
   240088    0.253    0.000    0.565    0.000 _weakrefset.py:58(__iter__)
   120056    0.091    0.000    0.134    0.000 _weakrefset.py:26(__exit__)
    30000    0.016    0.000    0.132    0.000 concurrent.py:259(add_done_callback)
    60000    0.060    0.000    0.128    0.000 gen.py:659(_contains_yieldpoint)
    30000    0.057    0.000    0.117    0.000 gen.py:822(callback)
    90000    0.058    0.000    0.095    0.000 functools.py:706(dispatch)
   120056    0.058    0.000    0.073    0.000 _weakrefset.py:20(__enter__)
420463/420200    0.061    0.000    0.064    0.000 {built-in method builtins.isinstance}
   120000    0.044    0.000    0.063    0.000 orderedweakset.py:17(__iter__)
    60000    0.025    0.000    0.059    0.000 {built-in method builtins.all}
    60004    0.028    0.000    0.053    0.000 {built-in method builtins.any}
    60000    0.027    0.000    0.050    0.000 concurrent.py:272(set_result)
   120056    0.043    0.000    0.043    0.000 _weakrefset.py:16(__init__)
    60000    0.039    0.000    0.039    0.000 concurrent.py:168(__init__)
    90000    0.036    0.000    0.036    0.000 weakref.py:393(__getitem__)
    90000    0.021    0.000    0.035    0.000 gen.py:815(<genexpr>)
   120056    0.032    0.000    0.032    0.000 _weakrefset.py:52(_commit_removals)
    60000    0.020    0.000    0.029    0.000 concurrent.py:355(is_future)
    60000    0.022    0.000    0.027    0.000 concurrent.py:335(__del__)
   120000    0.027    0.000    0.027    0.000 core.py:192(<listcomp>)
    90000    0.022    0.000    0.026    0.000 gen.py:668(<genexpr>)
    30000    0.016    0.000    0.024    0.000 concurrent.py:225(result)
    60000    0.023    0.000    0.023    0.000 concurrent.py:321(_set_done)
   120000    0.019    0.000    0.019    0.000 {built-in method builtins.iter}
   150112    0.018    0.000    0.018    0.000 {method 'add' of 'set' objects}
    60000    0.014    0.000    0.014    0.000 {method 'index' of 'tuple' objects}
   150256    0.014    0.000    0.014    0.000 {method 'remove' of 'set' objects}
    60056    0.012    0.000    0.012    0.000 {built-in method builtins.getattr}
    90006    0.012    0.000    0.012    0.000 {method 'append' of 'list' objects}
    60000    0.009    0.000    0.009    0.000 {method 'extend' of 'list' objects}
    30000    0.008    0.000    0.008    0.000 concurrent.py:219(_clear_tb_log)
```
Not the same exact timings but I think they are fairly close

**Comment by mrocklin** [comment]
Looks good enough to me.  I've rebased this into a couple commits and merged into master.  Closing.

**Comment by CJ-Wright** [comment]
@mrocklin Thank you very much!

**Comment by mrocklin** [comment]
Thanks for doing all the work CJ!  Sorry for the delay

**Pull request closed without merging**